src/storage/exec/IndexOutputNode.h

```diff
@@ -215,75 +198,152 @@ class IndexOutputNode final : public RelNode<T> {
+                }
+                v = std::move(retVal.value());
+            }
+        }
+
+        if (col.first) {
+            row.emplace_back(v);
+            dedup.emplace_back(std::move(v));
+        } else {
+            row.emplace_back(std::move(v));
+        }
+        return Status::OK();
+    }
+
+    Status addIndexValue(Row& row, Row& dedup, const kvstore::KV& data,
+                         const std::pair<bool, std::string>& col) {
+        nebula::Value v;
+        switch (QueryUtils::toReturnColType(col.second)) {
+            case QueryUtils::ReturnColType::kVid : {
+                auto vId = IndexKeyUtils::getIndexVertexID(planContext_->vIdLen_, data.first);
+                v = planContext_->isIntId_
+                    ? vId.toString()
+                    : vId.subpiece(0, vId.find_first_of('\0')).toString();
+                break;
+            }
+            case QueryUtils::ReturnColType::kTag : {
+                v = planContext_->tagId_;
+                break;
+            }
+            case QueryUtils::ReturnColType::kSrc : {
+                auto src = IndexKeyUtils::getIndexSrcId(planContext_->vIdLen_, data.first);
+                v = planContext_->isIntId_
+                    ? src.toString()
+                    : src.subpiece(0, src.find_first_of('\0')).toString();
+                break;
+            }
+            case QueryUtils::ReturnColType::kType : {
+                v = planContext_->edgeType_;
+                break;
+            }
+            case QueryUtils::ReturnColType::kRank : {
+                v = IndexKeyUtils::getIndexRank(planContext_->vIdLen_, data.first);
+                break;
+            }
+            case QueryUtils::ReturnColType::kDst : {
+                auto dst = IndexKeyUtils::getIndexDstId(planContext_->vIdLen_, data.first);
+                v = planContext_->isIntId_
+                    ? dst.toString()
+                    : dst.subpiece(0, dst.find_first_of('\0')).toString();
+                break;
+            }
+            default: {
+                v = IndexKeyUtils::getValueFromIndexKey(planContext_->vIdLen_,
+                                                        data.first,
+                                                        col.second,
+                                                        fields_,
+                                                        planContext_->isEdge_,
+                                                        hasNullableCol_);
+            }
+        }
+        if (col.first) {
+            row.emplace_back(v);
+            dedup.emplace_back(std::move(v));
         } else {
-            row.emplace_back(vId.subpiece(0, vId.find_first_of('\0')).toString());
+            row.emplace_back(std::move(v));
         }
+        return Status::OK();
     }
 
 private:
-    nebula::DataSet*                                  result_;
+    DeDupDataSet*                                     result_;
```

Comment:
How about move the logic of dedup to `dedup node`, I think we can just remain the output, otherwise, other method which will try to use dedup need to process like what you did here.

Reply:
> How about move the logic of dedup to `dedup node`, I think we can just remain the output, otherwise, other method which will try to use dedup need to process like what you did here.

Emm... Actually, I've thought about it . this might OOM more likely. because if just keep the single output, we need to de-dup the data via logic` vector -> map -> vector` for all of result . 
WDYT? and any good idea?

Reply:
Make sense for dedup, for now, the logic is output `map<row, row>` and convert to `vector<row>` in dedup node. What I concern is that other operator will consume more memory than dedup, such as `order by` and `group by`, so maybe we can ignore OOM here. This could be controlled with max count for temp solution if necessary. And maybe we need to add something like `memtracker` to control memory in global.

BTW, if storage can't dedup and have a OOM, graph will OOM as well.

Reply:
> Make sense for dedup, for now, the logic is output `map<row, row>` and convert to `vector<row>` in dedup node. What I concern is that other operator will consume more memory than dedup, such as `order by` and `group by`, so maybe we can ignore OOM here. This could be controlled with max count for temp solution if necessary. And maybe we need to add something like `memtracker` to control memory in global.
> 
> BTW, if storage can't dedup and have a OOM, graph will OOM as well.

Make sense.